File: src/CSharp/CodeCracker/Usage/MethodAnalyzers/MethodChecker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace CodeCracker.Usage.MethodAnalyzers
{
    public class MethodChecker
    {
        private readonly SyntaxNodeAnalysisContext context;
        private readonly DiagnosticDescriptor diagnosticDescriptor;

        public MethodChecker(SyntaxNodeAnalysisContext context, DiagnosticDescriptor diagnosticDescriptor)
        {
            this.context = context;
            this.diagnosticDescriptor = diagnosticDescriptor;
        }

        public void AnalyzeConstrutor(MethodInformation methodInformation)
        {
            if (ConstructorNameNotFound(methodInformation) || MethodFullNameNotFound(methodInformation.MethodFullDefinition))
            {
                return;
            }
            var argumentList = ((ObjectCreationExpressionSyntax)context.Node).ArgumentList;
            var arguments = GetArguments(argumentList);
            Execute(methodInformation, arguments, argumentList);
        }

        private bool ConstructorNameNotFound(MethodInformation methodInformation)
        {
            return AbreviatedConstructorNameNotFound(methodInformation) && QualifiedConstructorNameNotFound(methodInformation);
        }

        private bool AbreviatedConstructorNameNotFound(MethodInformation methodInformation)
        {
            var objectCreationExpressionSyntax = (ObjectCreationExpressionSyntax) context.Node;
            var identifier = objectCreationExpressionSyntax.Type as IdentifierNameSyntax;
            return identifier?.Identifier.ValueText != methodInformation.MethodName;
        }

        private bool QualifiedConstructorNameNotFound(MethodInformation methodInformation)
        {
            var objectCreationExpressionSyntax = (ObjectCreationExpressionSyntax)context.Node;
            var identifier = objectCreationExpressionSyntax.Type as QualifiedNameSyntax;
            return identifier?.Right.ToString() != methodInformation.MethodName;
        }

        public void AnalyzeMethod(MethodInformation methodInformation)
        {
            if (MethodNameNotFound(methodInformation) ||
                MethodFullNameNotFound(methodInformation.MethodFullDefinition))
            {
                return;
            }
            var argumentList = ((InvocationExpressionSyntax)context.Node).ArgumentList;
            var arguments = GetArguments(argumentList);

            Execute(methodInformation, arguments, argumentList);
        }

        private bool MethodNameNotFound(MethodInformation methodInformation)
        {
            var invocationExpression = (InvocationExpressionSyntax)context.Node;
            var memberExpression = invocationExpression.Expression as MemberAccessExpressionSyntax;
            return memberExpression?.Name?.Identifier.ValueText != methodInformation.MethodName;
        }

        private bool MethodFullNameNotFound(string methodDefinition)
        {
            var memberSymbol = context.SemanticModel.GetSymbolInfo(context.Node).Symbol;
            return memberSymbol?.ToString() != methodDefinition;
        }

        private void Execute(MethodInformation methodInformation, List<object> arguments, ArgumentListSyntax argumentList)
        {
            if (!argumentList.Arguments.Any())
            {
                return;
            }
            try
            {
                methodInformation.MethodAction.Invoke(arguments);
            }
            catch (Exception ex)
            {
                while (ex.InnerException != null)
                {
                    ex = ex.InnerException;
                }
                var diag = Diagnostic.Create(diagnosticDescriptor, argumentList.Arguments[methodInformation.ArgumentIndex].GetLocation(), ex.Message);
                context.ReportDiagnostic(diag);
            }
        }

        private List<object> GetArguments(ArgumentListSyntax argumentList)
        {
            return argumentList.Arguments
                .Select(a => a.Expression)
                .Select(l => l == null ? null : context.SemanticModel.GetConstantValue(l).Value)
                .ToList();
        }
    }
}

File: src/VisualBasic/CodeCracker/Usage/MethodAnalyzers/MethodChecker.vb
Imports CodeCracker.VisualBasic.Usage.MethodAnalyzers
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.Diagnostics
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Public Class MethodChecker

    Private ReadOnly _context As SyntaxNodeAnalysisContext
    Private ReadOnly _diagnosticDescriptor As DiagnosticDescriptor

    Public Sub New(context As SyntaxNodeAnalysisContext, diagnosticDescriptor As DiagnosticDescriptor)
        _context = context
        _diagnosticDescriptor = diagnosticDescriptor
    End Sub

    Public Sub AnalyzeConstructor(methodInformation As MethodInformation)
        If ConstructorNameNotFound(methodInformation) OrElse MethodFullNameNotFound(methodInformation.MethodFullDefinition) Then Exit Sub
        Dim argumentList = TryCast(_context.Node, ObjectCreationExpressionSyntax).ArgumentList
        Dim arguments = getarguments(argumentList)
        execute(methodInformation, arguments, argumentList)
    End Sub

    Private Function ConstructorNameNotFound(methodInformation As MethodInformation) As Boolean
        Return AbreviatedConstructorNameNotFound(methodInformation) AndAlso QualifiedConstructorNameNotFound(methodInformation)
    End Function

    Private Function AbreviatedConstructorNameNotFound(methodInformation As MethodInformation) As Boolean
        Dim objectCreationExpressionSyntax = DirectCast(_context.Node, ObjectCreationExpressionSyntax)
        Dim identifier = TryCast(objectCreationExpressionSyntax.Type, IdentifierNameSyntax)
        Return identifier?.Identifier.ValueText <> methodInformation.MethodName
    End Function

    Private Function QualifiedConstructorNameNotFound(methodInformation As MethodInformation) As Boolean
        Dim objectCreationExpressionSyntax = DirectCast(_context.Node, ObjectCreationExpressionSyntax)
        Dim identifier = TryCast(objectCreationExpressionSyntax.Type, QualifiedNameSyntax)
        Return identifier?.Right.ToString() <> methodInformation.MethodName
    End Function
    Public Sub AnalyzeMethod(methodInformation As MethodInformation)
        If MethodNameNotFound(methodInformation) OrElse MethodFullNameNotFound(methodInformation.MethodFullDefinition) Then
            Exit Sub
        End If
        Dim argumentList = DirectCast(_context.Node, InvocationExpressionSyntax).ArgumentList
        Dim arguments = GetArguments(argumentList)
        Execute(methodInformation, arguments, argumentList)
    End Sub

    Private Function MethodNameNotFound(methodInformation As MethodInformation) As Boolean
        Dim invocationExpression = DirectCast(_context.Node, InvocationExpressionSyntax)
        Dim memberExpression = TryCast(invocationExpression.Expression, MemberAccessExpressionSyntax)
        Return memberExpression?.name?.Identifier.ValueText <> methodInformation.MethodName
    End Function

    Private Function MethodFullNameNotFound(methodDefinition As String) As Boolean
        Dim memberSymbol = _context.SemanticModel.GetSymbolInfo(_context.Node).Symbol
        Return memberSymbol?.ToString <> methodDefinition
    End Function

    Private Sub Execute(methodInformation As MethodInformation, arguments As List(Of Object), argumentList As ArgumentListSyntax)
        If Not argumentList.Arguments.Any Then
            Exit Sub
        End If
        Try
            methodInformation.MethodAction.Invoke(arguments)
        Catch ex As Exception
            While (ex.InnerException IsNot Nothing)
                ex = ex.InnerException
            End While
            Dim diag = Diagnostic.Create(_diagnosticDescriptor, argumentList.Arguments(methodInformation.ArgumentIndex).GetLocation(), ex.Message)
            _context.ReportDiagnostic(diag)
        End Try
    End Sub

    Private Function GetArguments(argumentList As ArgumentListSyntax) As List(Of Object)
        Return argumentList.Arguments.
            Select(Function(a) a.GetExpression()).
            Select(Function(l) If(l Is Nothing, Nothing, _context.SemanticModel.GetConstantValue(l).Value)).
            ToList()
    End Function
End Class
